src/awkward/contents/indexedoptionarray.py

```diff
@@ -650,35 +647,38 @@ def _merging_strategy(self, others):
         head = [self]
         tail = []
 
-        i = 0
-        while i < len(others):
-            other = others[i]
+        it_others = iter(others)
+        for other in it_others:
             if isinstance(other, ak.contents.UnionArray):
+                tail.append(other)
+                tail.extend(it_others)
                 break
             else:
                 head.append(other)
-            i = i + 1
 
-        while i < len(others):
-            tail.append(others[i])
-            i = i + 1
-
-        if any(isinstance(x.backend.nplike, TypeTracer) for x in head + tail):
-            head = [
-                x if isinstance(x.backend.nplike, TypeTracer) else x.to_typetracer()
-                for x in head
-            ]
-            tail = [
-                x if isinstance(x.backend.nplike, TypeTracer) else x.to_typetracer()
-                for x in tail
-            ]
+        if any(x.backend.nplike.known_data for x in head + tail) and not all(
+            x.backend.nplike.known_data for x in head + tail
+        ):
+            raise RuntimeError
 
-        return (head, tail)
+        return head, tail
 
     def _reverse_merge(self, other):
         if isinstance(other, ak.contents.EmptyArray):
             return self
 
+        # FIXME: support categorical-categorical merging
+        if (
+            other.is_indexed
+            and other.parameter("__array__")
+            == self.parameter("__array__")
+            == "categorical"
+        ):
+            raise NotImplementedError(
+                "merging categorical arrays is currently not implemented. "
+                "Use `ak.enforce_type` to drop the categorical type and use general merging."
+            )
```

Comment:
_(I moved this comment because this is the relevant part of the code.)_

I was just talking with @lobis, and he has a use-case for `merge(categorical[X], categorical[X])` coming up soon (~month timescale). I'll be thinking about how to do this.

In the very first case (Luis's), the two categorical arrays have the same contents and no reindexing is necessary. We could kick this can down the road by only raising `NotImplementedError` when the contents of `categorical[X]` and `categorical[X]` are different. (And we only need to consider the case of `X = string`, since we don't know how to check equality for other types.)

Reply:
I suggest let's keep this as-is, but be thinking about next steps in anticipation for @lobis' work? Out-of-interest, which categoricals are we thinking about in that context? strings?

Reply:
https://github.com/lobis/geant4-python-application

The goal is to make a python interface for a generic Geant4 application which supplies Geant4 data as awkward arrays.

The structure of this data would mirror Geant4's Event > Track > Step hierarchy (the user would recieve an awkward array of events). Some of these fields are categorical strings: `track.particle_name`, `track.step.process`. `track.step.volume`, etc.

How it's currently laid out (very preliminary, @jpivarski and I already discussed some improvements to the awkward code). each Geant4 thread would produce an awkward array that would then be concatenated to the final array (this concatenation most likely taking place in python code).

> In the very first case (Luis's), the two categorical arrays have the same contents and no reindexing is necessary. We could kick this can down the road by only raising NotImplementedError when the contents of categorical[X] and categorical[X] are different. (And we only need to consider the case of X = string, since we don't know how to check equality for other types.)

I'm not sure if I'm following @jpivarski here. The categorical arrays may not have the same elements (most of them would be the same): some events may contain a very unlikely process that is not present in another batch of events, etc. Their indexing order could also be different as they could appear in different order.

Reply:
I was thinking of the physics processes, which I assume are enums in C++. You're right that there are also volumes, whose names are runtime-determined. Those could have different indexes that would have to get merged.

Reply:
> I was thinking of the physics processes, which I assume are enums in C++. You're right that there are also volumes, whose names are runtime-determined. Those could have different indexes that would have to get merged.

Physics process are also strings, there is no id-assignment (meaning they could get indexed differently). The only thing as far as I am aware that coul be considered an enum are the particles (PDG particle ID). In principle the full list of every categorical value should be available at the beginning of the simulation (list all available physics processes, particles, volumes, etc).